### DISCUSSION TITLE
How to get angle?

### DISCUSSION THREAD

**Discussion by Ken-Kuroki** [UTokyo-FieldPhenomics-Lab/EasyIDP#12] [Q&A] [answered]
Hi, I've been using EasyIDP to extract plots from drone images, and I wonder how to get the angles in which a given plot is captured in the images. I'm currently using the relative position of the plot within the image, and it works OK to pick the near-nadir images. But when you, for example, want to compare the accuracy of downstream analyses, the position in the image isn't the most intuitive metric to use, and I prefer to use angle instead. Do you have any suggestion to implement it using EasyIDP in a straightforward way? Thanks!

**Comment by HowcanoeWang** [comment]
Got a little confused about this description, do you mean the image rotation angle?
https://support.pix4d.com/hc/en-us/articles/202558969-Yaw-Pitch-Roll-and-Omega-Phi-Kappa-angles

If in this case, current EasyIDP doesn't read image rotation properties for pix4d and metashape. This may take some time to implement such function (the camera rotation in real-world != camera rotation in the reconstructed model, this need some convention)

---

Or you need something like this? (The green part angles?)

![image](https://user-images.githubusercontent.com/19968935/142967782-1680dff1-e602-4360-97a6-e7b4cd39ee4b.png)

**Comment by Ken-Kuroki** [comment, marked as answer]
My bad for unclear explanation. I think it's the latter. I want to get angles for 1, 2, and 3 in this drawing.
 
<img width="576" alt="image" src="https://user-images.githubusercontent.com/20122589/143042663-f567beda-199b-41f6-8c19-de487409d277.png">

**Reply by HowcanoeWang** [reply]
Please refer to this commit https://github.com/UTokyo-FieldPhenomics-Lab/EasyIDP/commit/51f0e74f189712be99957903ad5c30e48b7b60ce

You can get camera geo position by:
```python
# pix4d project
ximg_geo, yimg_geo, zimg_geo = p4d.img[img_name].cam_pos

# metashape project
# just coded by my mind, not 100% sure it works exactly, please let me know your progress
chunk_list = easyidp.io.metashape.open_project("your_metashape_path")
chunk0 = chunk_list[0]
ximg_geo, yimg_geo, zimg_geo = chunk0.photos[photo_id].get_camera_center()
```

You can get plot geo position by:
```python
xmin_geo, ymin_geo, zmin_geo = plot_geo.min(axis=0)
xmax_geo, ymax_geo, zmax_geo = plot_geo.max(axis=0)
xctr_geo = (xmax_geo + xmin_geo) / 2
yctr_geo = (ymax_geo + ymin_geo) / 2
zctr_geo = (zmax_geo + zmin_geo) / 2
```

Now you can calculate the unit vector to represent the angle between plot_center and img_center
https://stackoverflow.com/questions/52781433/how-to-get-the-unit-vector-from-a-numpy-array
```python
vector = np.asarray([ximg_geo, yimg_geo, zimg_geo]) - np.asarray([xctr_geo, yctr_geo, zctr_geo])
uvector = vector / (vector**2).sum()**0.5
```

**Reply by Ken-Kuroki** [reply]
Thanks for very very helpful response! It seems to be working for me.

**Reply by Ken-Kuroki** [reply]
Can I continue this topic with ver2.0 on Metashape projects?
To get camera coordinates, it seems I can use `ms._local2world(ms.photos[i].transform[0:3, 3])` to obtain world coordinates and then calculate distance from ROIs after converting their coordinates to world coordinate as well using `ms._crs2world(roi)`. However, it gives very large angles that seem incorrect. I think this is because the world coordinate system is, although I don't quite understand, the coordinate in the space not on the ground surface.

So I instead tried to convert photo coordinates to CRS with `ms._world2crs(ms._local2world(ms.photos[i].transform[0:3, 3]))`, which gave me the following error. It would be great if you can tell me how to avoid this error. If there are other packages to do the same conversion then that's just fine too.

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/root/.cache/pypoetry/virtualenvs/easyidptrial-xS3fZVNL-py3.7/lib/python3.7/site-packages/easyidp/metashape.py", line 99, in _world2crs
    return convert_proj3d(points_np, self.world_crs, self.crs)
  File "/root/.cache/pypoetry/virtualenvs/easyidptrial-xS3fZVNL-py3.7/lib/python3.7/site-packages/easyidp/metashape.py", line 1047, in convert_proj3d
    return out[0, :]
UnboundLocalError: local variable 'out' referenced before assignment
```

**Reply by HowcanoeWang** [reply]
I could replicate the error you mentioned here

```python
>>> import easyidp as idp
>>> lotus = idp.data.Lotus()
>>> ms = idp.Metashape(lotus.metashape.project, chunk_id=0) 
>>> ms._world2crs(ms._local2world(ms.photos[0].transform[0:3, 3])) 
array([139.54053245,  35.73458169, 130.09433649])
```
But if you changed the `ms.crs` to geotiff and then do the same thing:
```python
>>> import pyproj                                   
>>> ms.crs = pyproj.CRS.from_epsg(32654)              
>>> ms._world2crs(ms._local2world(ms.photos[0].transform[0:3, 3]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\OneDrive\Program\GitHub\EasyIDP\easyidp\metashape.py", line 99, in _world2crs
    return convert_proj3d(points_np, self.world_crs, self.crs)
  File "D:\OneDrive\Program\GitHub\EasyIDP\easyidp\metashape.py", line 1047, in convert_proj3d
    return out[0, :]
UnboundLocalError: local variable 'out' referenced before assignment
```
Fix this error info in 5913fdfc6d4a6b2c2f43a31640dc07e457efeaaa (at local, not pushed yet)

Temporary solution: 

Set `ms.crs = pyproj.CRS.from_epsg(4326)` first and then get photo location, and transform to GIS crs by 

```python
pos = {p.label: ms._world2crs(ms._local2world(p.transform[0:3, 3]))   for p in ms.photos}
transformer = pyproj.Transformer.from_crs("epsg:4326", "epsg:32654")

out = {}
for k,v in pos.items():
    # here the lon and lat need to be reversed for pyproj package
    transformed = transformer.transform(v[1], v[0], v[2])
    out[k] = np.asarray(transformed).T
```

Detailed explanation:
Ths crs in the Metashape is created from photos' meta_info, often epsg4326 with lon and lat, and the software will "display" in this coords. 

> local -> world -> crs(photo_meta)

When you export outputs from Metashape, you can specify another CRS for geotiff (in this case is epsg:32654)

> local -> world -> crs(photo_meta) -> geotiff

Probably in the next release, the `ms.crs` -> `ms.photo_meta_crs`, and then `ms.crs` -> geotiff.crs

and you can get photo position at current geotiff crs by 
```python
>>> ms.get_photo_position(photos[0], crs=geotiff.crs)
array([[ 368030.75661044, 3955476.94877388, 129.8508284]])
```

**Reply by Ken-Kuroki** [reply]
Thank you so much! I'll try that out!

**Reply by Ken-Kuroki** [reply]
It worked perfect! Appreciate your quick response.

**Reply by HowcanoeWang** [reply]
> It worked perfect! Appreciate your quick response.

In `v2.0.0.dev3`, you should be able to do:

```python
# get all photos and return a dict, and use ms.crs as output geo-coord
>>> ms.get_photo_position()   

# use given crs as output geo-coord
>>> ms.get_photo_position(to_crs=pyproj.CRS.from_epsg(4326))

# same as the pix4d
>>> p4d.get_photo_position(to_crs=pyproj.CRS.from_epsg(4326))
```
